File: python/src/brianterry_unicorn_maker/handlers.py
import logging


import requests
from cloudformation_cli_python_lib import (
    Action,
    OperationStatus,
    ProgressEvent,
    Resource,
    exceptions,
)

from .models import ResourceHandlerRequest, ResourceModel

# Use this logger to forward log messages to CloudWatch Logs.
LOG = logging.getLogger(__name__)
TYPE_NAME = "Brianterry::Unicorn::Maker"

resource = Resource(TYPE_NAME, ResourceModel)
test_entrypoint = resource.test_entrypoint

crud_crud_id = "<CRUDCRUD_API_ID_GOES_HERE>"
api_endpoint = f"https://crudcrud.com/api/{crud_crud_id}/unicorns"


def check_response(response, uid=None):
    if response.status_code == 404:
        raise exceptions.NotFound(TYPE_NAME, uid)
    if response.status_code not in [200, 201]:
        raise exceptions.InternalFailure(
            f"crudcrud.com error {response.status_code} {response.reason}"
        )



@resource.handler(Action.CREATE)
def create_handler(_, request: ResourceHandlerRequest, __) -> ProgressEvent:
    model = request.desiredResourceState
    progress: ProgressEvent = ProgressEvent(
        status=OperationStatus.SUCCESS, resourceModel=model,
    )
    response = requests.post(api_endpoint, json=model._serialize())
    check_response(response)
    progress.resourceModel.UID = response.json()["_id"]
    return progress


@resource.handler(Action.UPDATE)
def update_handler(_, request: ResourceHandlerRequest, __) -> ProgressEvent:
    model = request.desiredResourceState
    progress: ProgressEvent = ProgressEvent(
        status=OperationStatus.SUCCESS, resourceModel=model,
    )
    response = requests.put(f"{api_endpoint}/{model.UID}", json=model._serialize())
    check_response(response, model.UID)
    return progress


@resource.handler(Action.DELETE)
def delete_handler(_, request: ResourceHandlerRequest, __) -> ProgressEvent:
    model = request.desiredResourceState
    progress: ProgressEvent = ProgressEvent(
        status=OperationStatus.SUCCESS, resourceModel=model,
    )
    response = requests.delete(f"{api_endpoint}/{model.UID}")
    check_response(response, model.UID)
    return progress


@resource.handler(Action.READ)
def read_handler(_, request: ResourceHandlerRequest, __,) -> ProgressEvent:
    model = request.desiredResourceState
    response = requests.get(f"{api_endpoint}/{model.UID}")
    check_response(response, model.UID)
    model.Name = response.json()["Name"]
    model.Color = response.json()["Color"]
    return ProgressEvent(
        status=OperationStatus.SUCCESS,
        resourceModel=model,
    )


@resource.handler(Action.LIST)
def list_handler(_, __, ___) -> ProgressEvent:
    models = []
    response = requests.get(api_endpoint)
    check_response(response)
    for unicorn in response.json():
        models.append(ResourceModel(unicorn["_id"], unicorn["Name"], unicorn["Color"]))
    return ProgressEvent(
        status=OperationStatus.SUCCESS,
        resourceModels=models,
    )

File: typescript/src/handlers.ts
import {
    Action,
    BaseResource,
    exceptions,
    handlerEvent,
    HandlerErrorCode,
    LoggerProxy,
    OperationStatus,
    Optional,
    ProgressEvent,
    ResourceHandlerRequest,
    SessionProxy,
} from 'cfn-rpdk';
import fetch, { Response } from 'node-fetch';

import { ResourceModel } from './models';

interface Unicorn {
    readonly _id?: string;
    readonly Name?: string;
    readonly Color?: string;
}

class Resource extends BaseResource<ResourceModel> {
    // APIEndpoint is the crudcrud endpoint.
    // You can  obtain an endpoint by going to https://crudcrud.com.
    static readonly API_ENDPOINT = `https://crudcrud.com/api/<Your API ID>/unicorns`;
    static readonly DEFAULT_HEADERS = {
        'Accept': 'application/json; charset=utf-8',
        'Content-Type': 'application/json; charset=utf-8'
    };

    private async checkResponse(response: Response, logger: LoggerProxy, uid?: string): Promise<any> {
        if (response.status === 404) {
            throw new exceptions.NotFound(this.typeName, uid);
        } else if (![200, 201].includes(response.status)) {
            throw new exceptions.InternalFailure(
                `crudcrud.com error ${response.status} ${response.statusText}`,
                HandlerErrorCode.InternalFailure,
            );
        }
        const data = await response.text() || '{}';
        logger.log(`HTTP response ${data}`);
        return JSON.parse(data);
    }

    @handlerEvent(Action.Create)
    public async create(
        session: Optional<SessionProxy>,
        request: ResourceHandlerRequest<ResourceModel>,
        callbackContext: Map<string, any>,
        logger: LoggerProxy
    ): Promise<ProgressEvent<ResourceModel>> {
        logger.log('CREATE request', request);
        const model = new ResourceModel(request.desiredResourceState);
        if (model.id) {
            throw new exceptions.InvalidRequest('Create unicorn with readOnly property');
        }
        const progress = ProgressEvent.progress<ProgressEvent<ResourceModel>>(model);
        const body = model.toJSON();
        logger.log('UPDATE body', body);
        const response: Response = await fetch(Resource.API_ENDPOINT, {
            method: 'POST',
            headers: Resource.DEFAULT_HEADERS,
            body: JSON.stringify(body),
        });
        const jsonData: Unicorn = await this.checkResponse(response, logger);
        progress.resourceModel.id = jsonData._id;
        progress.status = OperationStatus.Success;
        logger.log('CREATE progress', progress);
        return progress;
    }

    @handlerEvent(Action.Update)
    public async update(
        session: Optional<SessionProxy>,
        request: ResourceHandlerRequest<ResourceModel>,
        callbackContext: Map<string, any>,
        logger: LoggerProxy
    ): Promise<ProgressEvent<ResourceModel>> {
        logger.log('UPDATE request', request);
        const model = new ResourceModel(request.desiredResourceState);
        const progress = ProgressEvent.progress<ProgressEvent<ResourceModel>>(model);
        const body = model.toJSON();
        delete body['Id'];
        logger.log('UPDATE body', body);
        const response: Response = await fetch(`${Resource.API_ENDPOINT}/${model.id}`, {
            method: 'PUT',
            headers: Resource.DEFAULT_HEADERS,
            body: JSON.stringify(body),
        });
        await this.checkResponse(response, logger, model.id);
        progress.status = OperationStatus.Success;
        logger.log('UPDATE progress', progress);
        return progress;
    }

    @handlerEvent(Action.Delete)
    public async delete(
        session: Optional<SessionProxy>,
        request: ResourceHandlerRequest<ResourceModel>,
        callbackContext: Map<string, any>,
        logger: LoggerProxy
    ): Promise<ProgressEvent<ResourceModel>> {
        logger.log('DELETE request', request);
        const model = new ResourceModel(request.desiredResourceState);
        const progress = ProgressEvent.progress<ProgressEvent<ResourceModel>>();
        const response: Response = await fetch(`${Resource.API_ENDPOINT}/${model.id}`, {
            method: 'DELETE',
            headers: Resource.DEFAULT_HEADERS,
        });
        await this.checkResponse(response, logger, model.id);
        progress.status = OperationStatus.Success;
        logger.log('DELETE progress', progress);
        return progress;
    }

    @handlerEvent(Action.Read)
    public async read(
        session: Optional<SessionProxy>,
        request: ResourceHandlerRequest<ResourceModel>,
        callbackContext: Map<string, any>,
        logger: LoggerProxy
    ): Promise<ProgressEvent<ResourceModel>> {
        logger.log('READ request', request);
        const model = new ResourceModel(request.desiredResourceState);
        const response: Response = await fetch(`${Resource.API_ENDPOINT}/${model.id}`, {
            method: 'GET',
            headers: Resource.DEFAULT_HEADERS,
        });
        const jsonData: Unicorn = await this.checkResponse(response, logger, model.id);
        model.name = jsonData.Name;
        model.color = jsonData.Color;
        const progress = ProgressEvent.success<ProgressEvent<ResourceModel>>(model);
        logger.log('READ progress', progress);
        return progress;
    }

    @handlerEvent(Action.List)
    public async list(
        session: Optional<SessionProxy>,
        request: ResourceHandlerRequest<ResourceModel>,
        callbackContext: Map<string, any>,
        logger: LoggerProxy
    ): Promise<ProgressEvent<ResourceModel>> {
        logger.log('LIST request', request);
        const response: Response = await fetch(Resource.API_ENDPOINT, {
            method: 'GET',
            headers: Resource.DEFAULT_HEADERS,
        });
        const jsonData: Unicorn[] = await this.checkResponse(response, logger);
        const progress = ProgressEvent.success<ProgressEvent<ResourceModel>>();
        progress.resourceModels = jsonData.map((unicorn: Unicorn) => {
            return new ResourceModel({
                id: unicorn._id,
                name: unicorn.Name,
                color: unicorn.Color,
            });
        });
        logger.log('LIST progress', progress);
        return progress;
    }
}

export const resource = new Resource(ResourceModel.TYPE_NAME, ResourceModel);

export const entrypoint = resource.entrypoint;

export const testEntrypoint = resource.testEntrypoint;
